### PR TITLE
Turrets target moving cars

### PR DESCRIPTION
**Pull request by joveeater**
#### Summary

SUMMARY: Features "Turrets, defense bots and mi-go now target moving cars"

#### Purpose of change

Make the tactic of avoiding enemies by crouching in your car less effective. More intelligent enemies (i.e. turrets, defense bots and mi-go scouts) will now target moving vehicles, favoring shooting at the controls if they can.

#### Describe the solution

There's a new json property in the gun special attack that controls the behavior. It causes the closest moving vehicle to be targeted if there's nothing better and it will favor shooting the controls of the vehicle if it can see them. Vehicles do not require any targeting time from the turret. 

#### Describe alternatives you've considered

Giving it to more monsters maybe? 

#### Testing

The old seems to work. 

#### Additional context

Crouching in your car is still a fairly effective strategy. It gives pretty good protection but it's a lot more risky. Note that this does work across z-levels so they will shoot at helis as they pass.

### CODE REVIEW

File: src/mattack_actors.cpp [hunk 1]
```diff
@@ -476,9 +480,26 @@ int gun_actor::get_max_range()  const
...
 }
 
+static vehicle *find_target_vehicle( monster &z )
+{
+    map &here = get_map();
+    vehicle *chosen = nullptr;
+    int distance = 60;
+    for( wrapped_vehicle &v : here.get_vehicles() ) {
+        int new_dist = rl_dist( z.pos(), v.pos );
+        if( v.v->velocity > 0 && new_dist < distance ) {
```
Comment (olanti-p) [inline src/mattack_actors.cpp:490]: IIRC vehicles that move backwards have negative velocity

**Comment by joveeater** [comment]
Good catch!

File: src/mattack_actors.cpp [hunk 2]
```diff
@@ -494,19 +515,38 @@ bool gun_actor::call( monster &z ) const
...
-            return false;
+            if( !target_moving_vehicles ) {
+                return false;
+            }
+            //No living targets, try to find a moving car
+            vehicle *veh = find_target_vehicle( z );
+            if( !veh ) {
+                return false;
+            }
+            for( const vpart_reference &vp : veh->get_avail_parts( "CONTROLS" ) ) {
```
Comment (Night-Pryanik) [inline src/mattack_actors.cpp:530]: I think it will make sense if they also can target wheels.

File: src/mattack_actors.cpp [hunk 3]
```diff
@@ -476,9 +480,26 @@ int gun_actor::get_max_range()  const
     return max_range;
 }
 
+static vehicle *find_target_vehicle( monster &z )
+{
+    map &here = get_map();
+    vehicle *chosen = nullptr;
+    int distance = 60;
```
Comment (Coolthulhu) [inline src/mattack_actors.cpp:487]: This should probably be an argument to the function, since we don't want to aim at things beyond range.

**Comment by olanti-p** [comment]
I think I found a bug, but I'm not sure whether it's caused by this PR or just by wacky monster targeting in general.
1. Spawn a vehicle, install `drive by wire controls` on it
2. Spawn a laser turret, stand just outside of its range
3. Use `remote vehicle controller` to drive the vehicle. The turret will ignore it as long as avatar stays visible.

**Comment by joveeater** [comment]
Yeah, this is a common thing for all monsters. If they can see the player they ignore everything else. I imagine it's the same if you throw zombies at it as well. Also lol at the turret struggling to stand.

**Comment by Coolthulhu** [comment]
Few notes:
* This allows exhausting turret ammo easily. Would be great if they had some sort of precalculation for viable targets, so that they won't shoot a moving wall of steel. Especially if they did start shooting its weak backside.
* It's OK to have the turret magically guess the controls actually used by the player rather than first-best.
* Use `VPFLAG_CONTROLS` over `"CONTROLS"` where possible, bitflag flags are faster to check.

**Pull request merged**